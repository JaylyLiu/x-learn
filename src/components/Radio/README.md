# Radio 单选框

## 使用
- 引用
```` jsx
import Radio, { RadioGroup } from '../italki-ui/Radio';
````

## Radio 用法
```` jsx
<Radio iistyle="success">Success</Radio>
````

## RadioGroup 用法
```` jsx
<RadioGroup name="radio" defaultValue="c">
  <Radio value="a">A</Radio>
  <Radio value="b">B</Radio>
  <Radio value="c">C</Radio>
  <Radio value="d">D</Radio>
</RadioGroup>
````
## API

### Radio 单选框

| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| iistyle | italki组件样式 | string | primary |
| autoFocus | 自动获取焦点 | boolean | false |
| checked | 指定当前是否选中 | boolean | false |
| defaultChecked | 初始是否选中 | boolean | false |
| value | 根据 value 进行比较，判断是否选中 | any | 无 |

### RadioGroup 

| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| defaultValue | 默认选中的value值 | string | primary |

## Functions

### Radio

| 名称 | 描述 |
| --- | --- |
| blur() | 移除焦点 |
| focus() | 获取焦点 |

### RadioGroup
| 名称 | 描述 |
| --- | --- |
| onChange() | 改变选中值 |
