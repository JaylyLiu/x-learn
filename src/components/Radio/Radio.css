/**
*  Radio style
*
**/

.i-radio-group {
  font-size: 14px;
  line-height: 1.5;
  color: rgba(0, 0, 0, 0.65);
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  list-style: none;
  display: inline-block;
  line-height: unset;
}

.i-radio-wrapper {
  margin: 0;
  margin-right: 8px;
}

.i-radio,
.i-radio-wrapper {
  font-size: 14px;
  line-height: 1.5;
  color: rgba(0, 0, 0, 0.65);
  box-sizing: border-box;
  padding: 0;
  list-style: none;
  display: inline-block;
  position: relative;
  white-space: nowrap;
  cursor: pointer;
}

.i-radio-wrapper-disabled {
  cursor: not-allowed;
}

.i-radio {
  margin: 0;
  outline: none;
  line-height: 1;
  vertical-align: text-bottom;
}

.i-radio-wrapper.primary .i-radio-focused .i-radio-inner,
.i-radio-wrapper.primary .i-radio-wrapper:hover .i-radio .i-radio-inner,
.i-radio-wrapper.primary .i-radio:hover .i-radio-inner {
  color: #ff4338;
}

.i-radio-wrapper.darken .i-radio-focused .i-radio-inner,
.i-radio-wrapper.darken .i-radio-wrapper:hover .i-radio .i-radio-inner,
.i-radio-wrapper.darken .i-radio:hover .i-radio-inner {
  color: #333;
}

.i-radio-wrapper.success .i-radio-focused .i-radio-inner,
.i-radio-wrapper.success .i-radio-wrapper:hover .i-radio .i-radio-inner,
.i-radio-wrapper.success .i-radio:hover .i-radio-inner {
  color: #5ebd5e;
}

.i-radio-checked:after {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 50%;
  content: '';
  animation: iRadioEffect 0.36s ease-in-out;
  animation-fill-mode: both;
  visibility: hidden;
  box-sizing: border-box;
}

.i-radio-wrapper.primary .i-radio-checked:after {
  border: 1px solid #ff4338;
}
.i-radio-wrapper.darken .i-radio-checked:after {
  border: 1px solid #333;
}
.i-radio-wrapper.success .i-radio-checked:after {
  border: 1px solid #5ebd5e;
}

.i-radio-wrapper:hover .i-radio:after,
.i-radio:hover:after {
  visibility: visible;
}

.i-radio-inner {
  position: relative;
  top: 0;
  left: 0;
  display: block;
  width: 18px;
  height: 18px;
  border-radius: 100px;
  box-sizing: border-box;
  border: 1px solid #d9d9d9;
  background-color: #fff;
  transition: all 0.3s;
}

.i-radio-inner:after {
  position: absolute;
  width: 8px;
  height: 8px;
  left: 4px;
  top: 4px;
  border-radius: 8px;
  display: table;
  border-top: 0;
  border-left: 0;
  content: ' ';
  opacity: 0;
  transform: scale(0);
  transition: all 0.3s cubic-bezier(0.78, 0.14, 0.15, 0.86);
}
.i-radio-wrapper.primary .i-radio-inner:after {
  background-color: #ff4338;
}

.i-radio-wrapper.darken .i-radio-inner:after {
  background-color: #333333;
}

.i-radio-wrapper.success .i-radio-inner:after {
  background-color: #5ebd5e;
}

.i-radio-input {
  position: absolute;
  left: 0;
  z-index: 1;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

.i-radio-wrapper.primary .i-radio-checked .i-radio-inner {
  border-color: #ff4338;
}

.i-radio-wrapper.darken .i-radio-checked .i-radio-inner {
  border-color: #333333;
}

.i-radio-wrapper.success .i-radio-checked .i-radio-inner {
  border-color: #5ebd5e;
}

.i-radio-checked .i-radio-inner:after {
  transform: scale(1);
  opacity: 1;
  transition: all 0.3s cubic-bezier(0.78, 0.14, 0.15, 0.86);
}

.i-radio-disabled .i-radio-inner {
  border-color: #d9d9d9 !important;
  background-color: #f5f5f5;
  cursor: not-allowed;
}

.i-radio-disabled .i-radio-inner:after {
  background-color: #ccc;
}

.i-radio-disabled .i-radio-input {
  cursor: not-allowed;
}

.i-radio-disabled + span {
  color: rgba(0, 0, 0, 0.25);
  cursor: not-allowed;
}

span.i-radio + * {
  padding-left: 8px;
  padding-right: 8px;
}

.i-radio-button-wrapper {
  margin: 0;
  height: 32px;
  line-height: 30px;
  color: rgba(0, 0, 0, 0.65);
  display: inline-block;
  transition: all 0.3s ease;
  cursor: pointer;
  border: 1px solid #d9d9d9;
  border-left: 0;
  border-top-width: 1.02px;
  background: #fff;
  padding: 0 15px;
  position: relative;
}

.i-radio-button-wrapper a {
  color: rgba(0, 0, 0, 0.65);
}

.i-radio-button-wrapper > .i-radio-button {
  margin-left: 0;
  display: block;
  width: 0;
  height: 0;
}

.i-radio-group-large .i-radio-button-wrapper {
  height: 40px;
  line-height: 38px;
  font-size: 16px;
}

.i-radio-group-small .i-radio-button-wrapper {
  height: 24px;
  line-height: 22px;
  padding: 0 7px;
}

.i-radio-button-wrapper:not(:first-child):before {
  content: '';
  display: block;
  top: 0;
  left: -1px;
  width: 1px;
  height: 100%;
  position: absolute;
  background-color: #d9d9d9;
}

.i-radio-button-wrapper:first-child {
  border-radius: 4px 0 0 4px;
  border-left: 1px solid #d9d9d9;
}

.i-radio-button-wrapper:last-child {
  border-radius: 0 4px 4px 0;
}

.i-radio-button-wrapper:first-child:last-child {
  border-radius: 4px;
}

.i-radio-button-wrapper-focused,
.i-radio-button-wrapper:hover {
  position: relative;
}

.i-radio-wrapper.primary .i-radio-button-wrapper-focused,
.i-radio-wrapper.primary .i-radio-button-wrapper:hover {
  color: #ff4338;
}

.i-radio-wrapper.darken .i-radio-button-wrapper-focused,
.i-radio-wrapper.darken .i-radio-button-wrapper:hover {
  color: #333;
}

.i-radio-wrapper.success .i-radio-button-wrapper-focused,
.i-radio-wrapper.success .i-radio-button-wrapper:hover {
  color: #5ebd5e;
}

/* .i-radio-button-wrapper .i-radio-inner,.i-radio-button-wrapper input[type=checkbox],.i-radio-button-wrapper input[type=radio] {
    opacity: 0;
    width: 0;
    height: 0
}

.i-radio-button-wrapper-checked {
    background: #fff;
    border-color: #1890ff;
    color: #1890ff;
    box-shadow: -1px 0 0 0 #1890ff;
    z-index: 1
}

.i-radio-button-wrapper-checked:before {
    background-color: #1890ff!important;
    opacity: .1
}

.i-radio-button-wrapper-checked:first-child {
    border-color: #1890ff;
    box-shadow: none!important
}

.i-radio-button-wrapper-checked:hover {
    border-color: #40a9ff;
    box-shadow: -1px 0 0 0 #40a9ff;
    color: #40a9ff
}

.i-radio-button-wrapper-checked:active {
    border-color: #096dd9;
    box-shadow: -1px 0 0 0 #096dd9;
    color: #096dd9
}

.i-radio-button-wrapper-disabled {
    cursor: not-allowed
}

.i-radio-button-wrapper-disabled,.i-radio-button-wrapper-disabled:first-child,.i-radio-button-wrapper-disabled:hover {
    border-color: #d9d9d9;
    background-color: #f5f5f5;
    color: rgba(0,0,0,.25)
}

.i-radio-button-wrapper-disabled:first-child {
    border-left-color: #d9d9d9
}

.i-radio-button-wrapper-disabled.i-radio-button-wrapper-checked {
    color: #fff;
    background-color: #e6e6e6;
    border-color: #d9d9d9;
    box-shadow: none
} */

@keyframes iRadioEffect {
  0% {
    transform: scale(1);
    opacity: 0.5;
  }

  to {
    transform: scale(1.6);
    opacity: 0;
  }
}
