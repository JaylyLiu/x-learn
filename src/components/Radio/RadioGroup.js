import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

class RadioGroup extends PureComponent {
  static childContextTypes = {
    radioGroup: PropTypes.any,
  };

  state = {
    value: this.props.defaultValue !== undefined ? this.props.defaultValue : '',
  };

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (
      !this.props.defaultValue &&
      nextProps.defaultValue !== undefined &&
      nextProps.defaultValue !== '' &&
      this.state.value === ''
    ) {
      this.setState({
        value: nextProps.defaultValue,
      });
      return;
    }

    if (nextProps.defaultValue !== this.props.defaultValue) {
      this.setState({
        value: nextProps.defaultValue,
      });
    }
  }

  getChildContext() {
    return {
      radioGroup: {
        onChange: this.onChange,
        value: this.state.value,
        disabled: this.props.disabled,
        name: this.props.name,
      },
    };
  }

  onChange = ev => {
    const lastValue = this.state.value;
    const { value } = ev.target;
    if (!('value' in this.props)) {
      this.setState({
        value,
      });
    }

    const onChange = this.props.onChange;
    if (onChange && value !== lastValue) {
      onChange(ev);
    }
  };

  render() {
    const { value } = this.state;
    const { children, className } = this.props;
    return (
      <div className={className}>
        {React.Children.map(
          children,
          (item, key) =>
            item &&
            React.cloneElement(item, {
              key,
              defaultChecked: value === item.props.value,
            })
        )}
      </div>
    );
  }
}

RadioGroup.childContextTypes = {
  radioGroup: PropTypes.any,
};

export default RadioGroup;
