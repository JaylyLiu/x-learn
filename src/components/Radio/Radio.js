import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import shallowEqual from 'shallowequal';
import RcCheckbox from 'rc-checkbox';
import './Radio.css';

class Radio extends Component {
  static propTypes = {
    iistyle: PropTypes.oneOf(['primary', 'darken', 'success']),
    defaultChecked: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),
    checked: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),
    disabled: PropTypes.bool,
    autoFocus: PropTypes.bool,
    value: PropTypes.any,
  };

  static defaultProps = {
    iistyle: 'primary',
    prefixCls: 'i-radio',
    type: 'radio',
  };

  static contextTypes = {
    radioGroup: PropTypes.any,
  };

  constructor(props) {
    super(props);
    // private ref
    this._rcCheckbox = null;
  }

  focus() {
    this._rcCheckbox.focus();
  }

  blur() {
    this._rcCheckbox.blur();
  }

  shouldComponentUpdate(nextProps, nextState, nextContext) {
    return (
      !shallowEqual(this.props, nextProps) ||
      !shallowEqual(this.state, nextState) ||
      !shallowEqual(this.context.radioGroup, nextContext.radioGroup)
    );
  }

  render() {
    const { context, props } = this;
    const {
      iistyle,
      prefixCls,
      className,
      contentClass,
      children,
      style,
      ...restProps
    } = props;
    const { radioGroup } = context;

    let radioProps = { ...restProps };
    if (radioGroup) {
      radioProps.name = radioGroup.name;
      radioProps.onChange = radioGroup.onChange;
      radioProps.checked = props.value === radioGroup.value;
      radioProps.disabled = props.disabled || radioGroup.disabled;
    }
    const wrapperClassString = cx(
      `${prefixCls}-wrapper`,
      iistyle,
      {
        [`${prefixCls}-wrapper-checked`]: radioProps.checked,
        [`${prefixCls}-wrapper-disabled`]: radioProps.disabled,
      },
      className
    );

    return (
      <label
        className={wrapperClassString}
        style={style}
        onMouseEnter={props.onMouseEnter}
        onMouseLeave={props.onMouseLeave}>
        <RcCheckbox
          {...radioProps}
          prefixCls={prefixCls}
          ref={node => {
            this._rcCheckbox = node;
          }}
        />
        {children !== undefined ? (
          <span className={contentClass}>{children}</span>
        ) : null}
      </label>
    );
  }
}

export default Radio;
