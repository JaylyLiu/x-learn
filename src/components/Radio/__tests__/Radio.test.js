import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Radio from '..';

describe('Radio', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Radio />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders default with label correctly', () => {
    const tree = renderer.create(<Radio>default</Radio>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders primary correctly', () => {
    const tree = renderer
      .create(<Radio iistyle="primary">primary</Radio>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders darken correctly', () => {
    const tree = renderer
      .create(<Radio iistyle="darken">darken</Radio>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders success correctly', () => {
    const tree = renderer
      .create(<Radio iistyle="success">success</Radio>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it('should work correctly with group', () => {
  // });
});
