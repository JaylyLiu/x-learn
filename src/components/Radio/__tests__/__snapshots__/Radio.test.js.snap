// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly 1`] = `
<label
  className="i-radio-wrapper primary"
>
  <span
    className="i-radio"
    style={Object {}}
  >
    <input
      checked={false}
      className="i-radio-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="i-radio-inner"
    />
  </span>
</label>
`;

exports[`Radio renders darken correctly 1`] = `
<label
  className="i-radio-wrapper darken"
>
  <span
    className="i-radio"
    style={Object {}}
  >
    <input
      checked={false}
      className="i-radio-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="i-radio-inner"
    />
  </span>
  <span>
    darken
  </span>
</label>
`;

exports[`Radio renders default with label correctly 1`] = `
<label
  className="i-radio-wrapper primary"
>
  <span
    className="i-radio"
    style={Object {}}
  >
    <input
      checked={false}
      className="i-radio-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="i-radio-inner"
    />
  </span>
  <span>
    default
  </span>
</label>
`;

exports[`Radio renders primary correctly 1`] = `
<label
  className="i-radio-wrapper primary"
>
  <span
    className="i-radio"
    style={Object {}}
  >
    <input
      checked={false}
      className="i-radio-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="i-radio-inner"
    />
  </span>
  <span>
    primary
  </span>
</label>
`;

exports[`Radio renders success correctly 1`] = `
<label
  className="i-radio-wrapper success"
>
  <span
    className="i-radio"
    style={Object {}}
  >
    <input
      checked={false}
      className="i-radio-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="i-radio-inner"
    />
  </span>
  <span>
    success
  </span>
</label>
`;
